src/v/resource_mgmt/memory_sampling.cc

```diff
@@ -77,66 +78,74 @@ void setup_additional_oom_diagnostics() {
+    // very quickly. An alternative approach that was tested was to hook this
+    // into the batch cache. However, doing this in reclaim is tricky as you
+    // want to avoid allocating there.  Equally, it's not entirely clear whether
+    // that is the right place as reclaim just "moves" memory from "reclaimable"
+    // to "free".
 
     size_t first_log_limit = _first_log_limit_fraction
                              * seastar::memory::stats().total_memory();
     size_t second_log_limit = _second_log_limit_fraction
                               * seastar::memory::stats().total_memory();
     size_t next_log_limit = first_log_limit;
 
-    while (true) {
-        try {
-            co_await _low_watermark_cond.wait([&next_log_limit]() {
-                auto current_low_water_mark
-                  = resources::available_memory::local()
-                      .available_low_water_mark();
-
-                return current_low_water_mark <= next_log_limit;
-            });
-        } catch (const ss::broken_condition_variable&) {
-            co_return;
-        }
-
-        auto allocation_sites = ss::memory::sampled_memory_profile();
-        const size_t top_n = std::min(size_t(5), allocation_sites.size());
-        top_n_allocation_sites(allocation_sites, top_n);
-
-        vlog(
-          _logger.info,
-          "{} {}",
-          diagnostics_header(),
-          fmt::join(
-            allocation_sites.begin(), allocation_sites.begin() + top_n, "|"));
-
-        if (next_log_limit == first_log_limit) {
-            next_log_limit = second_log_limit;
-        } else {
-            co_return;
-        }
-    }
+    _logging_timer.set_callback(
+      [this, first_log_limit, second_log_limit, next_log_limit]() mutable {
+          auto current_low_water_mark
+            = resources::available_memory::local().available_low_water_mark();
+
+          if (current_low_water_mark > next_log_limit) {
+              return;
+          }
+
+          auto allocation_sites = ss::memory::sampled_memory_profile();
+          const size_t top_n = std::min(size_t(5), allocation_sites.size());
+          top_n_allocation_sites(allocation_sites, top_n);
+
+          vlog(
+            _logger.info,
+            "{} bytes of available memory left - {} {}",
+            next_log_limit,
+            diagnostics_header(),
+            fmt::join(
+              allocation_sites.begin(), allocation_sites.begin() + top_n, "|"));
+
+          if (next_log_limit == first_log_limit) {
+              next_log_limit = second_log_limit;
+          } else {
+              _logging_timer.cancel();
+          }
+      });
+
+    _logging_timer.arm_periodic(_log_check_frequency);
 }
 
 memory_sampling::memory_sampling(
   ss::logger& logger, config::binding<bool> enabled)
-  : memory_sampling(logger, std::move(enabled), 0.2, 0.1) {}
+  : memory_sampling(
+    logger, std::move(enabled), std::chrono::seconds(60), 0.2, 0.1) {}
```

Comment:
nit: a smaller default may be okay since the callback is cheap in most cases and returns right away.

Reply:
Yeah it's not like there is a clear answer. We had a chat and Travis mentioned that all the cases where we were leaking over time we were leaking slowly so 60s should be enough.